File: src/algorithms/strings/group-anagrams/solution1.go
package main

import (
	"sort"
)

// O(w * n * log(n) + n * w * log(w)) itme | O(wn) space
// w = number of words, n = length of the longest word
func GroupAnagrams(words []string) [][]string {
	if len(words) == 0 {
		return [][]string{}
	}

	sortedWords := []string{}
	indices := []int{}
	for i, word := range words {
		sortedWords = append(sortedWords, sortWord(word))
		indices = append(indices, i)
	}
	sort.Slice(indices, func(i, j int) bool {
		return sortedWords[indices[i]] < sortedWords[indices[j]]
	})

	result := [][]string{}
	currentAnagramGroup := []string{}
	currentAnagram := sortedWords[indices[0]]
	for _, index := range indices {
		word := words[index]
		sortedWord := sortedWords[index]
		if len(currentAnagramGroup) == 0 {
			currentAnagramGroup = append(currentAnagramGroup, word)
			currentAnagram = sortedWord
			continue
		}

		if sortedWord == currentAnagram {
			currentAnagramGroup = append(currentAnagramGroup, word)
			continue
		}

		result = append(result, currentAnagramGroup)
		currentAnagramGroup = []string{word}
		currentAnagram = sortedWord
	}

	result = append(result, currentAnagramGroup)

	return result
}

func sortWord(word string) string {
	wordBytes := []byte(word)
	sort.Slice(wordBytes, func(i, j int) bool {
		return wordBytes[i] < wordBytes[j]
	})
	return string(wordBytes)
}

File: src/algorithms/strings/group-anagrams/solution1.js
// O(w * n * log(n)) + n * w * log(w) time | O(wn) space
// w = # of words
// n = length of the longest word
function groupAnagrams(words) {
	if (words.length === 0) return [];

	const sortedWords = words.map(word => word.split('').sort().join(''));
	const indices = [...Array(words.length).keys()];
	indices.sort((a, b) => {
		if (sortedWords[a] < sortedWords[b]) return -1;
		if (sortedWords[a] > sortedWords[b]) return 1;
		return 0;
	});

	const result = [];
	let currentAnagramGroup = [];
	let currentAnagram = sortedWords[indices[0]];
	for (const index of indices) {
		const word = words[index];
		const sortedWord = sortedWords[index];

		if (sortedWord === currentAnagram) {
			currentAnagramGroup.push(word);
			continue;
		};

		result.push(currentAnagramGroup);
		currentAnagramGroup = [word];
		currentAnagram = sortedWord;
	};

	result.push(currentAnagramGroup);

	return result;
};

exports.groupAnagrams = groupAnagrams;